SC_redesign/Crop_and_Soil/crop/leaf_area_index.py

```diff
@@ -31,30 +32,80 @@ def __init__(self):
+                                                                                    self._lai_shapes[1])
+
         self.determine_canopy_height()
-        self.check_for_senescence()
+        self.is_in_senescence = self.check_for_senescence
         if self.is_in_senescence:  # senescence
-            self.senesce_leaf_area()
+            self.leaf_area_index = self.determine_senescent_leaf_area_index(self.heat_fraction,
+                                                                            self.senescent_heat_fraction,
+                                                                            self.optimal_leaf_area_fraction)
         else:  # normal growth
             self.check_previous_leaf_area_values()
-            self.determine_max_leaf_area_change()
+            self.max_leaf_area_change = self.determine_max_leaf_area_change(self.optimal_leaf_area_fraction,
+                                                                            self.previous_optimal_leaf_area_fraction,
+                                                                            self.max_leaf_area_index,
+                                                                            self.previous_leaf_area_index)
             self.determine_leaf_area_added()
             self.add_leaf_area()
         self.shift_leaf_area_time()
 
-    def _determine_lai_shapes(self):
-        """sets lai shape parameters for optimal leaf area fraction calculations"""
-        self._lai_shapes = calc_shape_parameters(self.first_heat_fraction_point, self.second_heat_fraction_point,
-                                                 self.first_leaf_fraction_point, self.second_leaf_fraction_point)
+    @staticmethod
+    def determine_lai_shapes(first_heat_fraction: float, second_heat_fraction: float,
+                             first_leaf_fraction: float, second_leaf_fraction: float) -> List[float]:
+        """
+        calculates the shape coefficients for optimal LAI formula
+        replaces method calc_shape_parameters
+        """
+        if first_heat_fraction <= 0:
+            raise ValueError("first_heat_fraction must be greater than 0")
+        if second_heat_fraction <= 0:
+            raise ValueError("second_heat_fraction must be greater than 0")
+        if not 0 < first_leaf_fraction < 1:
+            raise ValueError("first_leaf_fraction must not be greater than 0 or less than 1")
+        if not 0 < second_leaf_fraction < 1:
+            raise ValueError("second_leaf_fraction must not be greater than 0 or less than 1")
+        if first_heat_fraction == second_heat_fraction:
+            # TODO: perhaps a way to handle this instead of throwing an error would be better
+            #   something like: second_heat_fraction += 1e-9
+            raise ValueError("first_heat_fraction cannot be exactly equal to second_heat_fractions")
+
+        # TODO: add error handling (see test_error_calc_shape_log() for conditions that fail)
+
```

Comment:
I'd like to update this todo
```suggestion
        # TODO: need to add any of these errors that get thrown when RuFaS runs to the  `OutputManager`. 
        #    This should probably be done in the `grow_canopy()` function 
        #    I'm still unsure how to do this effectively with warnings raised by static functions. - morrowcj
```